include/librepr/enum/util.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+#include <type_traits>
+#include <bit>
+#include <concepts>
+
+#include <librepr/ctvi/ctvi.h>
+#include <librepr/util/strings.h>
+
+namespace librepr::ctei {
+enum class EnumKind { Linear, Flags, Empty };
+
+template <typename T, EnumKind Kind>
+[[nodiscard]] constexpr auto to_underlying(auto idx) noexcept -> std::underlying_type_t<T> {
+  using underlying = std::underlying_type_t<T>;
+  if constexpr (Kind == EnumKind::Flags) {
+    if (idx == 0) {
+      return 0;
+    } else {
+      return underlying{1} << static_cast<underlying>(idx - 1);
+    }
+  } else {
+    return static_cast<underlying>(idx);
+  }
+}
+
+template <typename T, auto Value>
+[[nodiscard]] constexpr T to_enum() noexcept {
+  if constexpr (std::same_as<T, decltype(Value)>) {
+    return Value;
+  } else {
+    using underlying = std::underlying_type_t<T>;
+    // https://discourse.llvm.org/t/clang-16-notice-of-potentially-breaking-changes/65562
+    // GCC seems to warn for this as well
+    // TODO use this new diagnostic for search once https://github.com/llvm/llvm-project/issues/57176 is fixed
+
+    // library UB does not need to be diagnosed
+    // (https://eel.is/c++draft/expr.const#5.31) hence we can use std::bit_cast
+    // here to bypass the miscompile
+    // return __builtin_bit_cast(T, Value); //better debug output
```

Comment:
Why don't we use the builtin directly?

Reply:
pretty sure that's gcc/clang specific. There's no obvious performance impact from not using it directly and if it fails here we're in big trouble anyway :P